ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart2/exercises/1_CharacterizeAndRefactorE2EBackend/begin/tests/builders/assignmentBuilder.ts

```
@@ -0,0 +1,50 @@
+import { Assignment, Class } from '@prisma/client'
+import { prisma } from '../../src/database'
+
+interface IAssignmentBuilder {
+  fromClassRoom: (classRoom: Class) => IAssignmentBuilder
+  withTitle: (title: string) => IAssignmentBuilder 
+  build: () => Promise<Assignment | undefined>
+}
+
+export default function assignmentBuilder(): IAssignmentBuilder {
+  const state : {
+    classRoom: Class | null
+    title: string
+  } = {
+    classRoom: null,
+    title: ''
+  }
+  
+  const fromClassRoom = (classRoom: Class) => {
+    state.classRoom = classRoom
+    return builder
+  }
+  
+  const withTitle = (title: string) => {
+    state.title = title
+    return builder
+  }
+  
+  const build = async () => {
+    if (state.classRoom?.id) {
+      return await prisma.assignment.create({
+        data: {
+            classId: state.classRoom.id,
+            title: state.title
+        }
+    }); 
+    }
+  
+  return undefined
```

It's just my personal preference, but I'd let my builder to throw a Prisma error if classroom id is not defined. The tests would fail either way, but the return type of the function would be always an Assignment which would make it simpler IMO.

That way you could avoid optional chaining/undefined checks in your tests.

---

I didn't think about it that way. Thanks for the tip!